// @japa/core/build/src/Group/Runner.js
"use strict";
/*
 * @japa/core
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupRunner = void 0;
/**
 * Run all tests for a given group
 */
class GroupRunner {
    constructor(group, hooks, emitter) {
        this.group = group;
        this.hooks = hooks;
        this.emitter = emitter;
        /**
         * Reference to the startup runner
         */
        this.setupRunner = this.hooks.runner('setup');
        /**
         * Reference to the cleanup runner
         */
        this.teardownRunner = this.hooks.runner('teardown');
        /**
         * Test errors
         */
        this.errors = [];
        /**
         * Track if test has any errors
         */
        this.hasError = false;
    }
    /**
     * Notify the reporter about the group start
     */
    notifyStart() {
        const startOptions = { ...this.group.options };
        this.emitter.emit('group:start', startOptions);
    }
    /**
     * Notify the reporter about the group end
     */
    notifyEnd() {
        const endOptions = {
            ...this.group.options,
            hasError: this.hasError,
            errors: this.errors,
        };
        this.emitter.emit('group:end', endOptions);
    }
    /**
     * Running setup hooks
     */
    async runSetupHooks() {
        try {
            await this.setupRunner.run(this.group);
        }
        catch (error) {
            this.hasError = true;
            this.errors.push({ phase: 'setup', error });
        }
    }
    /**
     * Running teardown hooks
     */
    async runTeardownHooks() {
        try {
            await this.teardownRunner.run(this.group);
        }
        catch (error) {
            this.hasError = true;
            this.errors.push({ phase: 'teardown', error });
        }
    }
    /**
     * Running setup cleanup functions
     */
    async runSetupCleanupFunctions() {
        try {
            await this.setupRunner.cleanup(this.hasError, this.group);
        }
        catch (error) {
            this.hasError = true;
            this.errors.push({ phase: 'setup:cleanup', error });
        }
    }
    /**
     * Running teardown cleanup functions
     */
    async runTeardownCleanupFunctions() {
        try {
            await this.teardownRunner.cleanup(this.hasError, this.group);
        }
        catch (error) {
            this.hasError = true;
            this.errors.push({ phase: 'teardown:cleanup', error });
        }
    }
    /**
     * Run the test
     */
    async run() {
        this.notifyStart();
        /**
         * Run setup hooks and exit early when one of the hooks
         * fails
         */
        await this.runSetupHooks();
        if (this.hasError) {
            await this.runSetupCleanupFunctions();
            this.notifyEnd();
            return;
        }
        /**
         * Run the test executor
         */
        for (let test of this.group.tests) {
            await test.exec();
        }
        /**
         * Cleanup setup hooks
         */
        await this.runSetupCleanupFunctions();
        /**
         * Run + cleanup teardown hooks
         */
        await this.runTeardownHooks();
        await this.runTeardownCleanupFunctions();
        /**
         * Notify test end
         */
        this.notifyEnd();
    }
}
exports.GroupRunner = GroupRunner;
